### DISCUSSION TITLE
Version 1 release must haves

### DISCUSSION THREAD

**Discussion by jdwilkin4** [jdwilkin4/contrast-ratio-repo#94] [General]
This is a discussion thread to discuss what needs to be included in the version 1 release of this project. 

Here are my thoughts on what should be included:

- [ ] Hex code functionality and error messages shown for incorrect inputs
- [ ] named colors functionality and error messages shown for incorrect inputs 
- [ ] RGB functionality and error messages shown for incorrect inputs
- [ ] HSL functionality and error message shown for incorrect inputs
- [ ] HTML and CSS files that pass validators (https://validator.w3.org/)
- [ ] our project should meet basic accessibility standards and should not have large accessibility errors in it

**Comment by AndreyDmitriev39r** [comment]
I agree with all suggestions.

I would add functionality for named colors to that list. It's not hard to add it, as far as I figured, and I've already done some preparations. Refer to #79

Also, I would comment out or temporarily removed part of code for hsla and rgba for the first release. To make it work appears to be non-trivial task, requires more research. And many of apps which are similar to ours don't even bother with this.

If we'll be able to figure it out, we can add hsla/rgba stuff after release anyway.

And I think, since @DBenMoshe is already working on icons dynamic for graphics, we should add dynamic to the rest of icons.

And that would be it.

During working on the above we can at the same time polish/refactor codebase a little.

**Reply by jdwilkin4** [reply]
I have updated the list to include named colors as part of this first release. 
I am fine with leaving out the rgba and hsla for this release. 
 > And I think, since @DBenMoshe is already working on icons dynamic for graphics, we should add dynamic to the rest of icons.

We can create a new ticket specifically for the pass/fail icons

**Reply by AndreyDmitriev39r** [reply]
So, to wrap it all up a little

- ~~ticket for named colors contrast ratio already exists #79~~
  - UPDATE: this is merged 
- ticket for named colors invalid input could be created as soon as work on #79 finished
  - UPDATE: such issue can be created, though I have a little idea what would be the logic for this one: to spit some fun warning if user trying to eneter very long word? Not sure
- ticket about hsl invalid input could be created already
  - ~~UPDATE: I am planning to provide an opportunity for @dgupta0 to create this one. We have couple of examples of similar issues, so that would be good first experience in terms of creating an issue.~~
  - UPDATE2 we need to review latest version of @sparkO7 work for rgb error messsages and this stuff a bit complicated, so I am not considering this issue the beginner-friendly one to create right now
- ticket about pass/fail icons could be created already. Though we can wait for the draft from @DBenMoshe for #46 and see what logic is used there in order to follow somewhat similar logic about icons dynamic
- right before the release we can check console/validator and add tickets about according fixes if needed
- we should double-check if Bruce has more advices about accessibility at some point.
- Not sure how exactly we will deal with part of codebase about hsla/rgba. Removing? Commenting out? Leaving as it is?
  - ~~UPDATE: discussion about this one is going on below. As soon as we reach some consensus, issue about removing parts of code can be created~~
  - UPDATE Looks like we are going for removing option here, issue will be opened soon

## Major update 11 sept 2022

All of the above is kind of outdated now. The most important work we need to finish:

* issue #46
* issue #126
* issue #128
* check for console errors and for validator's errors

**Reply by jdwilkin4** [reply]
> UPDATE: such issue can be created, though I have a little idea what would be the logic for this one: to spit some fun warning if user trying to eneter very long word? 

I vote to keep it simple and just say "Sorry, that color is not in our database. Please try again". Or something to that affect. 

> and this stuff a bit complicated, so I am not considering this issue the beginner-friendly one to create right now

I think we should keep it as simple as possible.
For version 1, a user should be able to only input two valid rgb, hex, hsl, or named color values. 
I just tested the latest version of that PR and it looks like it is already doing that.
<img width="829" alt="Screen Shot 2022-09-01 at 12 45 17 PM" src="https://user-images.githubusercontent.com/67210629/187999295-4278ccbd-4376-41a7-95f3-d42c72d5329e.png">

Then for later versions we can implement more features to allow users to input different color format combinations. (Ex named colors and hex, rgb and hsl, etc)

We can change the line at the top to let users know what kinds of inputs we allow.

<img width="1050" alt="Screen Shot 2022-09-01 at 12 51 07 PM" src="https://user-images.githubusercontent.com/67210629/188000217-0474016a-8622-4bd6-9415-7a17c7959085.png">

I think we should consider what features absolutely need to be in this first release versus ones that can be added later on. 
It would be nice if the user could enter in different color format combinations but as we learned that requires a lot of extra validation and investigation into a clean approach. 
Does that have to be in this first release? 🤔 
That is something that the team has to decide 👍

**Reply by AndreyDmitriev39r** [reply]
I agree that we don't need feature for different color formats input in first release. In my experiments about that(I mentioned them in discord) I wanted to check how easy it can be done - to provide ratio for any color pair. And it appears it is not super complex, but still it will be serious refactoring. So, yes, this is not what we need for first release in my opinion.

> We can change the line at the top to let users know what kinds of inputs we allow.

Oh, this should definitely be done for the first release I think.

**Comment by Okibe00** [comment]
>Not sure how exactly we will deal with part of codebase about hsla/rgba. Removing? Commenting out? Leaving as it is?  

I think we should comment it out, since we will be working on it for V2 we can just wrap it within a comment that says `V2 feature ` or something.

**Reply by AndreyDmitriev39r** [reply]
That would be one option.
But we can always extract this code from commit history, it doesn't need to be placed in actual codebase.

https://github.com/DBenMoshe/contrast-ratio-repo/commit/1a51a24ba7d754fed1bb19f6d44976b70ffc8526

**Reply by Okibe00** [reply]
Ok, this problem is settled then.
